packages/walletconnect-v2/src/index.ts

```diff
@@ -86,27 +102,54 @@ export class WalletConnect extends Connector {
     this.events.emit(URI_AVAILABLE, uri)
   }
 
+  private async initializeProvider(
+    desiredChainId: number | undefined = this.defaultChainId
+  ): Promise<WalletConnectProvider> {
+    const ethProviderModule = await import('@walletconnect/ethereum-provider')
+    const rpcMap = this.rpcMap ? await getBestUrlMap(this.rpcMap, this.timeout) : undefined
+    const chainProps = this.getChainProps(desiredChainId)
+
+    this.provider = await ethProviderModule.default.init({
+      ...this.options,
+      ...chainProps,
+      rpcMap,
+    })
+
+    return this.handleProviderEvents(this.provider)
+  }
+
+  // Helper function to reorder chains array based on desiredChainId
+  private reorderChainsBasedOnId(
+    chains: number[] | undefined,
+    desiredChainId: number | undefined
+  ): number[] | undefined {
+    if (!chains || !desiredChainId || !chains.includes(desiredChainId) || chains.length === 0) {
+      return chains
+    }
+    return getChainsWithDefault(chains, desiredChainId)
+  }
+
+  private getChainProps(desiredChainId: number | undefined = this.defaultChainId): ChainsProps {
+    // Reorder chains and optionalChains if necessary
+    const orderedChains = this.reorderChainsBasedOnId(this.options.chains, desiredChainId)
+    const orderedOptionalChains = this.reorderChainsBasedOnId(this.options.optionalChains, desiredChainId)
+
+    // Validate and return the result
+    if (isArrayOneOrMore(orderedChains)) {
+      return { chains: orderedChains, optionalChains: orderedOptionalChains }
+    } else if (isArrayOneOrMore(orderedOptionalChains)) {
```

Comment:
What is the difference between this branch of code and previous branch of code? Looks like you're literally executing same code. Could this be (if necessary) inlined into a single conditional, such as `if (isArrayOneOrMore(orderedChains) || isArrayOneOrMore(orderedOptionalChains)`?

Reply:
unfortunately this is not compatible with typescript's type discrimination system.

```
    if (isArrayOneOrMore(orderedChains) || isArrayOneOrMore(orderedOptionalChains)) {
      return { chains: orderedChains, optionalChains: orderedOptionalChains }
    }
```
gives the pictured error

<img width="1092" alt="image" src="https://github.com/Uniswap/web3-react/assets/5773490/70082595-8b62-4b02-b4a0-4d989f360651">